Client/game_sa/CWorldSA.cpp

```diff
@@ -350,6 +388,157 @@ bool CWorldSA::ProcessLineOfSight(const CVector* vecStart, const CVector* vecEnd
+            CVector             hitPosOS{};             //< Hit position in object space
+            CEntitySAInterface* entity{};               //< The hit entity
+        } c = {};
+
+        c.entity = targetEntity;
+
+        // Get matrix, and it's inverse
+        targetEntity->Placeable.matrix->ConvertToMatrix(c.entMat);
+        c.entInvMat = c.entMat.Inverse();
+
+        // ...to transform the line origin and end into object space
+        c.originOS     = c.entInvMat.TransformVector(*vecStart);
+        c.endOS        = c.entInvMat.TransformVector(*vecEnd);
+        c.dirOS        = c.endOS - c.originOS;
+        c.minHitDistSq = c.dirOS.LengthSquared(); // By setting it to this value we avoid collisions that would be detected after the line segment [but are still ont the ray]
+
+        //const auto lineMagSq = (ctx.originOS - ctx.endOS).LengthSquared();
+
+        // Do raycast against the DFF to get hit position material UV and name
+        // This is very slow
+        // Perhaps we could paralellize it somehow? [OpenMP?]
+        const auto ProcessOneAtomic = [](RpAtomic* a, void* data) {
+            const auto c = (Context*)data;
+            const auto f = RpAtomicGetFrame(a);
+
+            const auto GetFrameCMatrix = [](RwFrame* f) {
+                CMatrix out;
+                pGame->GetRenderWare()->RwMatrixToCMatrix(*RwFrameGetMatrix(f), out);
+                return out;
+            };
+
+            // Atomic not visible
+            if (!a->renderCallback || !(a->object.object.flags & 0x04 /*rpATOMICRENDER*/)) {
+                return true;
+            }
+
+            // Sometimes atomics have no geometry [I don't think that should be possible, but okay]
+            const auto geo = a->geometry;
+            if (!geo) {
+                return true;
+            }
+
+            // Calculate transformation by traversing the hierarchy from the bottom (this frame) -> top (root frame)
+            CMatrix localToObjTransform{};
+            for (auto i = f; i && i != i->root; i = RwFrameGetParent(i)) {
+                localToObjTransform = GetFrameCMatrix(i) * localToObjTransform;
+            }
+            const auto objToLocalTransform = localToObjTransform.Inverse();
+
+            const auto ObjectToLocalSpace = [&](const CVector& in) {
+                return objToLocalTransform.TransformVector(in);
+            };
+
+            // Transform from object space, into local (the frame's) space
+            const auto localOrigin = ObjectToLocalSpace(c->originOS);
+            const auto localEnd = ObjectToLocalSpace(c->endOS);
+
+#if 0
+            if (!CCollisionSA::TestLineSphere(
+                CColLineSA{localOrigin, 0.f, localEnd, 0.f},
+                reinterpret_cast<CColSphereSA&>(*RpAtomicGetBoundingSphere(a)) // Fine for now
+            )) {
+                return true; // Line segment doesn't touch bsp
+            }
+#endif
```

Comment:
Is there any reason to keep this code?

Reply:
Yeah, hopefully, eventually, it'll be fixed